### PR TITLE
CLN: clean color selection in _matplotlib/style

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

There was a comment in the top of the module that the code was a bit too dynamic.
The logic in the function was too complex.

I refactored the code
 - Extract functions
 - Simplify logic
 - Add type annotations

Note: there is one issue with mypy, which I do not know how to handle. Currently I suggested to ignore the error.

### CODE REVIEW

File: pandas/plotting/_matplotlib/style.py [hunk 1]
```diff
@@ -1,81 +1,163 @@
...
+        else:
+            return list(color)
+
+    # ignoring mypy error here
+    # error: Argument 1 to "list" has incompatible type
+    # "Union[Sequence[float], Collection[Union[str, Sequence[float]]]]";
+    # expected "Iterable[Union[str, Sequence[float]]]"  [arg-type]
+    # A this point color may be sequence of floats or series of colors,
+    # all convertible to list
+    return list(color)  # type: ignore [arg-type]
```
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/style.py:122]: emm, this seems to be a bit more complex than the original comprehension, is there a reason to write this way?

```python
list(color)	
if is_list_like(color) and not isinstance(color, dict)	
else color
```

File: pandas/plotting/_matplotlib/style.py [hunk 2]
```diff
@@ -1,81 +1,163 @@
...
+
+
+def _get_colors_from_color_type(color_type: str, num_colors: int) -> Collection[Color]:
```
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/style.py:125]: List should be enough

File: pandas/plotting/_matplotlib/style.py [hunk 3]
```diff
@@ -1,81 +1,163 @@
...
+    """Get colors from user input color type."""
+    if color_type == "default":
+        return _get_default_colors(num_colors)
+    elif color_type == "random":
+        return _get_random_colors(num_colors)
+    else:
+        raise ValueError("color_type must be either 'default' or 'random'")
+
+
+def _get_default_colors(num_colors: int) -> Collection[Color]:
```
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/style.py:135]: List should be enough

File: pandas/plotting/_matplotlib/style.py [hunk 4]
```diff
@@ -1,81 +1,163 @@
...
+    """Get ``num_colors`` of default colors from matplotlib rc params."""
+    import matplotlib.pyplot as plt
+
+    # need to call list() on the result to copy so we don't
+    # modify the global rcParams below
+    try:
+        colors = [c["color"] for c in list(plt.rcParams["axes.prop_cycle"])]
+    except KeyError:
+        colors = list(plt.rcParams.get("axes.color_cycle", list("bgrcmyk")))
+
```
Comment (charlesdong1991) [inline pandas/plotting/_matplotlib/style.py:145]: in the original code, seems there is a check of `colors` in this branch, so does it mean after this refactoring, this won't happen? or `colors` shouldn't be a str in the first place once reaching this part of code?
```python
if isinstance(colors, str):	
     colors = list(colors)
```

**Review by charlesdong1991 — CHANGES_REQUESTED** [review]
thanks @ivanovmg 

emm, maybe @simonjayhawkins can provide some feedbacks on the mypy error here?

File: pandas/plotting/_matplotlib/style.py [hunk 5]
```diff
@@ -1,81 +1,163 @@
...
         colors += multiple * colors
         colors += colors[:mod]
 
     return colors
 
 
+def _get_colors_from_colormap(
+    colormap: Union[str, "Colormap"],
+    num_colors: int,
+) -> Collection[Color]:
```
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/style.py:83]: can use Collection for arguments to be permissive where appropriate, but return type should be more precise, in this case List.

File: pandas/plotting/_matplotlib/style.py (hunk 1 above)
Comment (simonjayhawkins) [inline pandas/plotting/_matplotlib/style.py:122]: the ci failure could be fixed with 
```suggestion
    return list(color)  # type: ignore[arg-type]
```
but please avoid using ignore instead.

File: pandas/plotting/_matplotlib/style.py [hunk 6]
```diff
@@ -1,81 +1,163 @@
...
+    )
+
+    if isinstance(colors, dict):
+        return colors
+
+    return _cycle_colors(list(colors), num_colors=num_colors)
+
+
+def _get_colors(
+    *,
```
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:46]: Just to enforce keyword-only arguments.

File: pandas/plotting/_matplotlib/style.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:83]: OK, will adjust for return types. I remember exactly @simonjayhawkins suggesting to use more generic types. Thus, I used collection.

File: pandas/plotting/_matplotlib/style.py [hunk 7]
```diff
@@ -1,81 +1,163 @@
...
+            raise ValueError(f"Colormap {cmap} is not recognized")
+    return colormap
+
+
+def _get_colors_from_color(
+    color: Union[Color, Dict[str, Color], Collection[Color]],
+) -> Union[Dict[str, Color], Collection[Color]]:
+    """Get colors from user input color."""
+    if isinstance(color, Iterable) and len(color) == 0:
+        raise ValueError("Invalid color argument: {color}")
```
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:104]: I added this check to avoid try/except ZeroDivisionError (in the original code), which would be happening when color == "". I will make it an f-string, I missed that.

File: pandas/plotting/_matplotlib/style.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:122]: The original portion of the code was exactly quite dynamic. I decided to make it more explicit. I now understand that I missed one case here (which does not hit in tests). This case is when color is sequence of floats (apparently the internal code calls get_standard_colors only with color being sequence of sequence of floats). I will add this edge case and the corresponding test. Presumably, mypy errors will also be resolved.

File: pandas/plotting/_matplotlib/style.py (hunk 4 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:145]: Right, in this particular function colors will not be a string. This check is effectively carried out in ``_get_colors_from_color``.

File: pandas/plotting/_matplotlib/style.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py]: Honestly speaking, I kind of liked the original comprehension. The problem is that it was not compatible with typing.

**Comment by ivanovmg** [comment]
Here is a problem. I want to install Cycler (https://matplotlib.org/cycler/) for testing purposes, to mimic matplotlib behavior on default color selection. When I add it into ``environment.yml``, it is rejected by pre-commit check. Moreover, it seems that this package is never installed anyway, which leads to the failure of the tests requiring Cycler.

**Comment by ivanovmg** [comment]
I have some weird problem with the new test module ``test_style.py``. It works fine in my local virtual environment, but fails in CI (some importing problem).

```
_____________ ERROR collecting pandas/tests/plotting/test_style.py _____________
ImportError while importing test module '/home/vsts/work/1/s/pandas/tests/plotting/test_style.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniconda3/envs/pandas-dev/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pandas/tests/plotting/test_style.py:5: in <module>
    from pandas.plotting._matplotlib.style import get_standard_colors
pandas/plotting/_matplotlib/__init__.py:3: in <module>
    from pandas.plotting._matplotlib.boxplot import (
pandas/plotting/_matplotlib/boxplot.py:5: in <module>
    from matplotlib.artist import setp
E   ModuleNotFoundError: No module named 'matplotlib'
```

**Comment by ivanovmg** [comment]
All checks are good except Windows py38_np18, where tests related to numba failed. I guess it is irrelevant of my changes.

File: pandas/plotting/_matplotlib/style.py [hunk 8]
```diff
@@ -9,74 +19,156 @@
...
+        return _get_colors_from_color_type(color_type, num_colors=num_colors)
 
-    if isinstance(colors, str) and _is_single_color(colors):
-        # GH #36972
-        colors = [colors]
 
-    # Append more colors by cycling if there is not enough color.
-    # Extra colors will be ignored by matplotlib if there are more colors
-    # than needed and nothing needs to be done here.
+def _cycle_colors(colors: List[Color], num_colors: int) -> List[Color]:
```
Comment (jreback) [inline pandas/plotting/_matplotlib/style.py:68]: is there a reason we don't just use cycler?

**Comment by jreback** [comment]
cc @charlesdong1991

File: pandas/plotting/_matplotlib/style.py (hunk 8 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:68]: Good question. Probably we can use it. I will check.

**Comment by ivanovmg** [comment]
I checked that we actually can use cycler.
But the problem is that tests must be changed.
Currently it is acceptable to return more colors that ``num_colors``.

When I use cycler I would prefer to keep only required number of colors:

```
def _cycle_colors(colors: List[Color], num_colors: int) -> List[Color]:
    import matplotlib.pyplot as plt

    cycle = plt.cycler(color=colors)
    for _, item in zip(range(num_colors), cycle()):
        yield item["color"]

```

This looks simple to understand, IMHO.
And it seems more reasonable that we return exactly ``num_colors`` colors, without any extra (although extra colors are ignored anyway).
But are we good that I go ahead and change some tests?
Or maybe leave it to another PR?

File: pandas/plotting/_matplotlib/style.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/style.py]: yep. that's correct for function arguments. return types should be precise.

File: pandas/tests/plotting/test_style.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+import pandas.util._test_decorators as td
+
+from pandas import Series
+
+with suppress(ImportError):
+    from pandas.plotting._matplotlib.style import get_standard_colors
+
+
+@td.skip_if_no_mpl
```
Comment (simonjayhawkins) [inline pandas/tests/plotting/test_style.py:13]: you could just use pytest.importorskip see pandas\tests\plotting\test_converter.py or other test modules testing optional dependencies. e.g. pytables

Comment (ivanovmg) [reply pandas/tests/plotting/test_style.py]: Did that. Please check if this is what you suggested.

File: pandas/tests/plotting/test_style.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        "num_colors, expected",
+        [
+            (1, ["b"]),
+            (3, ["b", "g", "r"]),
+            (4, ["b", "g", "r", "y"]),
+            (5, ["b", "g", "r", "y", "b"]),
+            (7, ["b", "g", "r", "y", "b", "g", "r"]),
+        ],
+    )
+    def test_default_colors_named_from_prop_cycle_string(self, num_colors, expected):
```
Comment (charlesdong1991) [inline pandas/tests/plotting/test_style.py:41]: i _think_ you could combine this one with the test above, and add `cycler_color` to parameterize since those two tests look almost identical?

File: pandas/tests/plotting/test_style.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+    @pytest.mark.parametrize(
+        "num_colors, expected",
+        [
+            (1, ["r", "g", "b", "k"]),
+            (2, ["r", "g", "b", "k"]),
+            (3, ["r", "g", "b", "k"]),
+            (4, ["r", "g", "b", "k"]),
+            (5, ["r", "g", "b", "k", "r"]),
+            (6, ["r", "g", "b", "k", "r", "g"]),
```
Comment (charlesdong1991) [inline pandas/tests/plotting/test_style.py:107]: i _think_ it's nicer to just test the case where num_color </>/= color, so 3 parametrizations is enough. no need to change here though ^^

**Review by charlesdong1991 — APPROVED** [review]
sorry for late response, i was fallen ill in the past week. 

@ivanovmg  looks very good to me and original logics seem to be covered and unchanged! nice job!

only two minor suggestions on the tests to make the tests slightly shorter, also okay if not change.

File: pandas/plotting/_matplotlib/style.py [hunk 12]
```diff
@@ -9,74 +19,156 @@
...
+    """Get colors from user input color."""
+    if len(color) == 0:
+        raise ValueError(f"Invalid color argument: {color}")
+
+    if isinstance(color, str) and _is_single_color(color):
+        # GH #36972
+        return [color]
+
+    if _is_floats_color(color):
+        color = cast(Sequence[float], color)
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/style.py:114]: This cast seems unnecessary - is there an error this solves?

File: pandas/plotting/_matplotlib/style.py [hunk 13]
```diff
@@ -9,74 +19,156 @@
...
+        return [color]
+
+    color = cast(Collection[Color], color)
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/style.py:117]: This cast also seems to be hiding a lot of other issues - does this generate an error when removed? `_is_floats_color` imposes a few more conditions than just Sequence versus Collection

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/plotting/_matplotlib/style.py (hunk 12 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:114]: Yes, cast is necessary.
At this point ``color`` is of type ``Union[Color, Collection[Color]]``. And for some reason ``_is_floats_color`` check does not filter out ``Collection[Color]``. So, instead of ignoring, I added cast here.

File: pandas/plotting/_matplotlib/style.py (hunk 13 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:117]: Removing this cast does result in error:
```
pandas\plotting\_matplotlib\style.py:117: error: Argument 1 to "_gen_list_of_colors_from_iterable" has incompatible type "Union[Union[str, Sequence[float]], Collection[Union[str, Sequence[float]]]]"; expected "Collection[Union[str, Sequence[float]]]"  [arg-type]
```

Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:117]: It is expected that ``_is_floats_color`` will select only those matching type ``Sequence[float]``.
But probably mypy could not comprehend the logic in ``_is_floats_color``.
So it would consider that at the moment of return statement ``color`` variable can be still of type ``Union[Color, Collection[Color]]``, however, in fact it would be only ``Collection[Color]`` after passing through two preceding if statements.

File: pandas/plotting/_matplotlib/style.py (hunk 12 above)
Comment (WillAyd) [reply pandas/plotting/_matplotlib/style.py:114]: Functions can't narrow types yet in mypy, but regardless this is pretty confusing. Can you try to refactor this to make things clearer?

Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:114]: I do not know how to make it even clearer.
The logic is the following.

1. Start with color being maybe
 a. a single string color
 b. or multiple string color (like 'rgbk')
 c.  or single float color (0.1, 0.2, 0.3)
 d. or multiple float colors
2. Address option a.
3. Address option b.
4. Address options c and d.

Is the logic confusing or the two cast statements are?

Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py]: @WillAyd, I updated the logic a little bit.
1. Address one color (either string or sequence of floats)
2. Address collection of colors

Casting is still there as there are no other options.

Please let me know what you think about this implementation.

File: pandas/plotting/_matplotlib/style.py (hunk 8 above)
Comment (jreback) [reply pandas/plotting/_matplotlib/style.py:68]: can you update here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. a couple of comments (aside from @WillAyd and @charlesdong1991 )

File: pandas/plotting/_matplotlib/style.py (hunk 8 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:68]: Oh yes, please check my comment https://github.com/pandas-dev/pandas/pull/37203#issuecomment-713441530

File: pandas/plotting/_matplotlib/style.py [hunk 14]
```diff
@@ -9,74 +19,174 @@
...
 def get_standard_colors(
-    num_colors: int, colormap=None, color_type: str = "default", color=None
+    num_colors: int,
+    colormap: Optional["Colormap"] = None,
+    color_type: str = "default",
+    color: Optional[Union[Dict[str, Color], Color, Collection[Color]]] = None,
 ):
-    import matplotlib.pyplot as plt
-
+    if isinstance(color, dict):
```
Comment (jreback) [inline pandas/plotting/_matplotlib/style.py:35]: can you add a doc-strin ghere

File: pandas/plotting/_matplotlib/style.py [hunk 15]
```diff
@@ -9,74 +19,174 @@
...
+
+
+def _get_colors(
+    *,
+    color: Optional[Union[Color, Collection[Color]]],
+    colormap: Optional[Union[str, "Colormap"]],
+    color_type: str,
+    num_colors: int,
+) -> List[Color]:
+    """Get colors from user input."""
```
Comment (jreback) [inline pandas/plotting/_matplotlib/style.py:55]: can you add a doc-string. 

also the current summary is not very  descriptive, nor is the function name .

File: pandas/plotting/_matplotlib/style.py (hunk 8 above)
Comment (jreback) [reply pandas/plotting/_matplotlib/style.py:68]: ok cool. pls make an issue about this / separate followon PR. I *would* actually like to use cycler if we can (it post-dates a lot of this code).

**Review by jreback — CHANGES_REQUESTED** [review]
just a few minor comments, ping on green.

File: pandas/plotting/_matplotlib/style.py (hunk 14 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py:67]: Added. Green. Please take a look.
@jreback

File: pandas/plotting/_matplotlib/style.py (hunk 15 above)
Comment (ivanovmg) [reply pandas/plotting/_matplotlib/style.py]: I changed it into ``_derive_colors``. I am not sure if you would consider it a better alternative. I thought of having a more verbose name (like ``_derive_colors_from_cmap_color_and_type``, but that does not seem reasonable to me, although I added the explanation in the docstring.
Anyway derive seems a better verb, implying some logic underneath.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ivanovmg very nice
